### ISSUE TITLE
How to log hparams to Tensorboard?

### ISSUE THREAD

**Issue by RafailFridman** [PyTorchLightning/pytorch-lightning#1225]
Hello! I'm trying to view my hparams on tensorboard, but can't actually see them there. As I understood from documentation, to log hparams one should add self.hparams in the __init__ of the LightningModule. Here's what I'm doing:
```python
class MyModule(pl.LightningModule):
    def __init__(self,hparams):
        super().__init__()
        self.hparams =  Namespace(**{'learning_rate': hparams.learning_rate,
                                    'batch_size':hparams.batch_size,
                                    'normalize_embeddings':hparams.normalize_embeddings,
                                    'normalize': hparams.normalize,
                                    'k_neighbors':hparams.k_neighbors,
                                    'melspec_dir':hparams.melspec_dir})
```
My hparams also contain Trainer hparams, so to log only the right ones, I wrote this workaround. But anyway, I could not see any of my hparams. What could be the problem?

**Comment by mRcSchwering** [comment]
Hi,

I am trying to figure the same thing out since 2 days now.
So, for just writing the hparams to tensorboard, you can do this:
```
class MyModule(pl.LightningModule):

    def __init__(self, hparams: dict):
        super(MyModule, self).__init__()
        self.hparams = hparams

```
It seems that the `logger.log_hyperparams` is called with those `self.hparams` before the training begins.

**Comment by mRcSchwering** [comment]
However, I have no idea what to do to add metrics (_e.g._ validation loss of best epoch) to a set of hyperparameters. Can i extend the question to that?

Currently, I am changing the logger like this:
```
class MyTensorBoardLogger(TensorBoardLogger):

    def __init__(self, *args, **kwargs):
        super(MyTensorBoardLogger, self).__init__(*args, **kwargs)

    def log_hyperparams(self, *args, **kwargs):
        pass

    @rank_zero_only
    def log_hyperparams_metrics(self, params: dict, metrics: dict) -> None:
        params = self._convert_params(params)
        exp, ssi, sei = hparams(params, metrics)
        writer = self.experiment._get_file_writer()
        writer.add_summary(exp)
        writer.add_summary(ssi)
        writer.add_summary(sei)
        # some alternative should be added
        self.tags.update(params)
```
And then I'm writing the hparams with metrics in a callback:
```
def on_train_end(self, trainer, module):
        module.logger.log_hyperparams_metrics(module.hparams, {'val_loss': self.best_val_loss})
```
But that doesn't seem right. 
Is there a way to write some metric together with the hparams as well?

**Comment by RafailFridman** [comment]
Thanks for the reply!
> Did you try defining your `Namespace` outside, then passing it as `hparams`?

hparams, that I pass to __init__ of my module is already a Namespace that contains many hyperparameters, including those that I want to log in tensorboard.

**Comment by mRcSchwering** [comment]
I never tried it with a namespace, I always use dicts.
Did you try it with a dict?

**Comment by RafailFridman** [comment]
No, I didn't
According to #651 hparams should be a Namespace object, so I assume that will not work, but I'll check.

**Comment by mRcSchwering** [comment]
#1228 is not related

**Comment by RafailFridman** [comment]
Do you know when `self.logger.log_hyperparams(hparams)` is called? Does it happen at the end of training or every epoch?

**Comment by mRcSchwering** [comment]
when you overwrite it with a `print` you can see that it gets called at the very beginning of the training (not training step or epoch, the whole training).

**Comment by RafailFridman** [comment]
I'm still not able to log any hparams to Tensorboard, although I exactly follow the documentation and do this:
```python
class MyModule(pl.LightningModule):
  def __init__(self, hparams):
    super().__init__()
    self.hparams = hparams
```
,where hparams, that I'm passing to init are `parser.parse_args()`
Does anybody have similar problems?

**Comment by RafailFridman** [comment]
I tried explicitly add hparams to tb like this:
```python
 self.logger.experiment.add_hparams(hparam_dict= 
                                            {'learning_rate': self.hparams.learning_rate},
                                           metric_dict = dict())
```
But that didn't work.

**Comment by RafailFridman** [comment]
Ok, I it. It looks like that TensorBoard won't display hparams if some previous runs were without them. So I had to start a new tb process with a clean logs directory. 
Everything works normal now.

**Comment by lkhphuc** [comment]
Oh Jesus, I've been wasting so much time debugging Lightning. Thank @RafailFridman .

**Comment by Skyy93** [comment]
@Borda 

Excuse me, could you please mention this behaviour in the docs? Thank you

**Comment by Borda** [comment]
> Excuse me, could you please mention this behaviour in the docs? Thank you

That sounds as a good idea, mind send a PR with suggestion?
